### PR TITLE
Expose `DeviceProxy::GetAttestationChallenge` to Obj-C.

### PR DESCRIPTION
**Pull request by mburshteyn1**
#### Problem
* Expose `DeviceCommissioner::GetAttestationChallenge` to Obj-C.

#### Change overview
Added function `getAttestationChallengeForDeviceId` to `MTRDeviceController` which exposes the `GetAttestationChallenge` from `DeviceProxy`

### CODE REVIEW

**Review by mspang — APPROVED** [review]

File: src/darwin/Framework/CHIP/MTRDeviceController.h [hunk 1]
```diff
@@ -130,6 +130,9 @@ typedef void (^MTRDeviceConnectionCallback)(MTRBaseDevice * _Nullable device, NS
  */
 - (void)setNocChainIssuer:(id<MTRNOCChainIssuer>)nocChainIssuer queue:(dispatch_queue_t)queue;
 
+/** Return the attestation challenge for the secure session of the device being commissioned. */
+- (nullable NSData *)getAttestationChallengeForDeviceId:(uint64_t)deviceId;
```
Comment (woody-apple) [inline src/darwin/Framework/CHIP/MTRDeviceController.h:134]:
```suggestion
- (nullable NSData *)generateAttestationChallengeForDeviceId:(uint64_t)deviceId;
```

File: src/darwin/Framework/CHIP/MTRDeviceController.mm [hunk 2]
```diff
@@ -668,6 +668,37 @@ - (void)setNocChainIssuer:(id<MTRNOCChainIssuer>)nocChainIssuer queue:(dispatch_
     });
 }
 
+- (nullable NSData *)getAttestationChallengeForDeviceId:(uint64_t)deviceId
```
Comment (woody-apple) [inline src/darwin/Framework/CHIP/MTRDeviceController.mm:671]:
```suggestion
- (nullable NSData *)generateAttestationChallengeForDeviceId:(uint64_t)deviceId
```

**Review by woody-apple — APPROVED** [review]

File: src/darwin/Framework/CHIP/MTRDeviceController.mm [hunk 3]
```diff
@@ -703,6 +703,37 @@ - (nullable NSData *)computePaseVerifier:(uint32_t)setupPincode iterations:(uint
     return result;
 }
 
+- (nullable NSData *)generateAttestationChallengeForDeviceId:(uint64_t)deviceId
+{
+    __block CHIP_ERROR errorCode = CHIP_ERROR_INCORRECT_STATE;
+    if (![self isRunning]) {
+        [self checkForError:errorCode logMsg:kErrorNotRunning error:nil];
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/MTRDeviceController.mm:710]: Please use `checkIsRunning`.

File: src/darwin/Framework/CHIP/MTRDeviceController.mm [hunk 4]
```diff
@@ -703,6 +703,37 @@ - (nullable NSData *)computePaseVerifier:(uint32_t)setupPincode iterations:(uint
...
+        return nil;
+    }
+
+    __block NSData * attestationChallenge;
+    dispatch_sync(_chipWorkQueue, ^{
+        if ([self isRunning]) {
+            chip::CommissioneeDeviceProxy * deviceProxy;
+            errorCode = self.cppCommissioner->GetDeviceBeingCommissioned(deviceId, &deviceProxy);
+            if (errorCode != CHIP_NO_ERROR) {
+                [self checkForError:errorCode logMsg:@"Invalid Attestation Challenge device ID." error:nil];
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/MTRDeviceController.mm:720]: The point of `checkForError` is that it returns whether there was an error, so can be:
```
   if ([self checkForError:errorCode ...]) {
       return;
   }
```
etc.

File: src/darwin/Framework/CHIP/MTRDeviceController.mm [hunk 5]
```diff
@@ -703,6 +703,37 @@ - (nullable NSData *)computePaseVerifier:(uint32_t)setupPincode iterations:(uint
...
+                return;
+            }
+
+            NSMutableData * challengeBuffer = [[NSMutableData alloc] initWithLength:chip::Crypto::kAES_CCM128_Key_Length];
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/MTRDeviceController.mm:724]: This is a 16-byte buffer.  Just put it on the stack, please:
```
    uint8_t challengeBuffer[chip::Crypto::kAES_CCM128_Key_Length];
    chip::ByteSpan challenge(challengeBuffer);
```

File: src/darwin/Framework/CHIP/MTRDeviceController.mm [hunk 6]
```diff
@@ -703,6 +703,37 @@ - (nullable NSData *)computePaseVerifier:(uint32_t)setupPincode iterations:(uint
...
+            chip::ByteSpan challenge((uint8_t *) [challengeBuffer mutableBytes], chip::Crypto::kAES_CCM128_Key_Length);
+
+            errorCode = deviceProxy->GetAttestationChallenge(challenge);
+            MTR_LOG_ERROR("GetAttestationChallenge: %s", chip::ErrorStr(errorCode));
+
+            attestationChallenge = [NSData dataWithBytes:challenge.data() length:challenge.size()];
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/MTRDeviceController.mm:730]:
```suggestion
            attestationChallenge = AsData(challenge);
```

File: src/darwin/Framework/CHIP/MTRDeviceController.mm [hunk 7]
```diff
@@ -703,6 +703,37 @@ - (nullable NSData *)computePaseVerifier:(uint32_t)setupPincode iterations:(uint
...
+            if (errorCode != CHIP_NO_ERROR) {
+                [self checkForError:errorCode logMsg:@"Invalid Attestation Challenge device ID." error:nil];
+                return;
+            }
+
+            NSMutableData * challengeBuffer = [[NSMutableData alloc] initWithLength:chip::Crypto::kAES_CCM128_Key_Length];
+            chip::ByteSpan challenge((uint8_t *) [challengeBuffer mutableBytes], chip::Crypto::kAES_CCM128_Key_Length);
+
+            errorCode = deviceProxy->GetAttestationChallenge(challenge);
+            MTR_LOG_ERROR("GetAttestationChallenge: %s", chip::ErrorStr(errorCode));
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/MTRDeviceController.mm:728]: But then we go ahead and return the garbage value in `challenge`?  Does not look right.

File: src/darwin/Framework/CHIP/MTRDeviceController.h [hunk 8]
```diff
@@ -134,6 +134,9 @@ typedef void (^MTRDeviceConnectionCallback)(MTRBaseDevice * _Nullable device, NS
  */
 - (void)setNocChainIssuer:(id<MTRNOCChainIssuer>)nocChainIssuer queue:(dispatch_queue_t)queue;
 
+/** Return the attestation challenge for the secure session of the device being commissioned. */
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/MTRDeviceController.h:137]: Please document under what conditions this returns nil.

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

File: src/darwin/Framework/CHIP/MTRDeviceController.mm [hunk 9]
```diff
@@ -703,6 +703,37 @@ - (nullable NSData *)computePaseVerifier:(uint32_t)setupPincode iterations:(uint
     return result;
 }
 
+- (nullable NSData *)generateAttestationChallengeForDeviceId:(uint64_t)deviceId
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/MTRDeviceController.mm:706]: Oh, and I know @woody-apple asked for this to be called "generate..." but that does not seem right.  The attestation challenge either exists or not; it was generated during session setup.  This is just getting its value if it does exist.  @woody-apple given that, what would you suggest as the name?

Comment (mburshteyn1) [reply src/darwin/Framework/CHIP/MTRDeviceController.mm:708]: I appreciate the thorough review @bzbarsky-apple. @woody-apple please let me know your thoughts on naming.

Comment (woody-apple) [reply src/darwin/Framework/CHIP/MTRDeviceController.mm:708]: I am fine with anything, let’s get this in now, and fix naming in our API review.

File: src/darwin/Framework/CHIP/MTRDeviceController.h [hunk 10]
```diff
@@ -134,6 +134,14 @@ typedef void (^MTRDeviceConnectionCallback)(MTRBaseDevice * _Nullable device, NS
  */
 - (void)setNocChainIssuer:(id<MTRNOCChainIssuer>)nocChainIssuer queue:(dispatch_queue_t)queue;
 
+/**
+ * Return the attestation challenge for the secure session of the device being commissioned.
+ *
+ * Attempts to retreive the generated attestation challenge from a commissionee with the given Device ID.
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/MTRDeviceController.h:140]:
```suggestion
 * Attempts to retrieve the generated attestation challenge from a commissionee with the given Device ID.
```

**Review by bzbarsky-apple — APPROVED** [review]

**Comment by andy31415** [comment]
Added to 1.0 because platform-specific

**Pull request merged**